### ISSUE TITLE
[BACK-END] Save-In-Progress Bug (Inflection Header)

### ISSUE THREAD

**Issue by edmangimelli** [department-of-veterans-affairs/va.gov-team#17595]
This is the back-end ticket. [Here's the front-end ticket](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17411)

# Problem

Olive Branch is butchering some keys in our _save-in-progress_ data

--specifically, keys that have hyphens in them:
```
  "treated_disability_names": {
    "Some text here ... Type-1 Diabetes": true
  }
```

Why are some keys full sentences (like the `Some text here..` key above)? For form526, a Veteran selects/inputs their disabilities and then, later in the form, those disabilities get tied to the VA Facility that treated that disability. When connecting those disabilities to a VA Facility, they become checkboxes --something like:

```
VA Facility: Bob Dole Medical Center

Choose which disabilties relate to that period of service:

- [ ] Some text here ... Type-1 Diabetes
- [ ] Some more text here ... PTSD 
- [ ] Back injury ... more text ... vertebrae C-4


VA Facility: Maryland National Guard Hospital

Choose which disabilties relate to that period of service:

- [ ] Some text here ... Type-1 Diabetes
- [ ] Some more text here ... PTSD 
- [ ] Back injury ... more text ... vertebrae C-4

```

The front-end form system encodes those checkboxes as:

```
"disability string": boolean
```
So something like:
```
"Lower back injury affecting vertebrae C-4 vertebrae C-5": true
```

The back-end uses a middleware (Olivebranch), that invisibily changes the case of JSON keys between requests.

 - The front-end sends camel-case, the back-end get snake-case auto-magically.
 - The front-end asks for the data back, the back-end sends snake-case _but_ the front-end receives camel-case.

This works fine in most scenarios, but, sometimes, this happens:

```
irb(main):009:0>  "theQuickBrownFox".underscore.camelize(:lower)
=> "theQuickBrownFox"

irb(main):010:0> "Hello there, Sam-I-Am!".underscore.camelize(:lower)
=> "hello there, samIAm!"
```
See how in the second example the hyphens disappeared? (also the capitalization)

Why does it matter if this `Hello there, Sam-I-Am!` becomes `hello there, samIAm!`?

Two reasons:

1) For our form, the checkbox `hello there, samIAm!: true`, won't match any known disabilities (which will be in an array in a different part of the SIP data).

2) EVSS is expecting the disability string to be exactly how it was originally formatted --they won't know what to do with _`for "VA Medical Facility A", disability "hello there, samIAm!" applies`_

# Possible Solutions

1. Stop doing the automatic case-conversion completely (at least for save-in-progress routes)
2. Only convert the keys for part of the payload
3. Tweak back-end plaftorm code so that we can optionally ignore Olivebranch for certain forms
4. Tweak front-end plaftorm code so that a form can optionally opt-out of Olivebranch (opt-out of sending the inflection header)

### 1. Stop doing the automatic case-conversion completely

The point of Olivebranch is to allow the front end and back end to be blissfully unaware that the other stores data in a different format. The front end sends camel-case and gets back camel-case --it doesn't seem like any transformation ever happened.

Then, theoretically, turning off Olivebranch shouldn't change anything. Well --it shouldn't change the _form data_ part of the payload. In a perfect world, the front end would store data in the back end, and the back end would give it back ***without ever looking at the data***, but I believe there are some metadata fields that the back end ***does*** look at. The back end _would_ need to start expecting those fields to be camel-case.

_Also_, if there is any existing workaround code in the front end, it might break if suddenly the SIP keys stop being transformed.

### 2. Only convert the keys for part of the payload

This would require tweaking the Olivebranch gem or doing something equally hacky. [Philip believes this is a bad route to go,](https://dsva.slack.com/archives/CBU0KDSB1/p1606941146437600?thread_ts=1606932092.415400&cid=CBU0KDSB1) and I agree.

### 3. Tweak back-end plaftorm code so that we can optionally ignore Olivebranch ...

I think this is completely possible. Olive branch let's you ignore certain routes. But... ⬇

### 4. Tweak front-end plaftorm code so that a form can optionally opt-out of Olivebranch (opt-out of sending the inflection header)

I think the best option is to _not_ send the inflection header (sometimes), as opposed to ignoring it.

# The Next Problems!

A. What happens to SIP data during the changeover (i.e. it gets saved pre-solution, and retrieved post-solution).

B. How do we fix the currently borked SIP data we have in the DB.

Those are roughly the same problem, and I believe BAM team (VSA Benefits and Memorials Team) can work on figuring out how to fix our borked data separately from the platform problem above.

# Conclusion

Can we spin up a platform ticket to tackle this problem? I don't have a number, but I suspect the impact to users is not insignificant.

I will change [issue 17411](https://github.com/department-of-veterans-affairs/va.gov-team/issues/17411
) into the BAM ticket for fixing existing data in the DB (_The Next Problems!_ section).
# Links

https://github.com/department-of-veterans-affairs/va.gov-team/issues/15368

https://github.com/department-of-veterans-affairs/va.gov-team/issues/15310

https://dsva.slack.com/archives/CBU0KDSB1/p1606932092415400

https://github.com/department-of-veterans-affairs/vsp-support/issues/46

https://github.com/department-of-veterans-affairs/va.gov-team/issues/17411

# Acceptance Criteria
(this ticket has been converted from being a writeup of the problem to being the back-end implementation ticket)

- [x] make is so that the back-end (InProgressForm controller/model) can handle ***not*** receiving the inflection header

**Comment by luke-gcio** [comment]
Added it to our backlog although it may end off moving to another team's backlog.

**Comment by edmangimelli** [comment]
[I have a PR for this](https://github.com/department-of-veterans-affairs/vets-api/pull/5582)

**Comment by edmangimelli** [comment]
Tweaking my PR for [Anna's feedback](https://github.com/department-of-veterans-affairs/vets-api/pull/5582#issuecomment-752756244)

**Comment by edmangimelli** [comment]
brainstormed with Anna about _the next problems_ (see "The Next Problems!" A & B above).
We've decided that fixing all of the data in the DB with a rake task if safer than doing it on the fly. She's spinning up a PR for that rake task as I finish my PR.

Working on some more of her feedback.

Spinning up a new PR for adding some InProgressFormSerializer tests (I'll merge that PR first). **I want to be absolutely certain that my changes to the serializer don't change any existing behavior.** My PR is effectively a _platform code_ change, and I don't want to inadvertently change the format of our serialization for any form.

**Comment by edmangimelli** [comment]
PR https://github.com/department-of-veterans-affairs/vets-api/pull/5582

**Comment by edmangimelli** [comment]
responded to feedback. tweaks made. PR needing review (posted in vfs-platform-support)

**Comment by edmangimelli** [comment]
made tweaks for @erluti feedback

**Comment by edmangimelli** [comment]
still waiting on a review for this one

**Comment by edmangimelli** [comment]
updating the documentation

**Comment by edmangimelli** [comment]
followup PRs:
https://github.com/department-of-veterans-affairs/vets-api/pull/5662
https://github.com/department-of-veterans-affairs/vets-api/pull/5671
https://github.com/department-of-veterans-affairs/vets-api/pull/5678

**Comment by edmangimelli** [comment]
[main PR](https://github.com/department-of-veterans-affairs/vets-api/pull/5582) merged and deployed 🎉
